Lib/idlelib/idle_test/test_calltip.py

```diff
@@ -88,10 +88,7 @@ class SB:  __call__ = None
 replacement repl.  repl can be either a string or a callable;
 if a string, backslash escapes in it are processed.  If it is
 a callable, it's passed the Match object and must return''')
-        tiptest(p.sub, '''\
-(repl, string, count=0)
-Return the string obtained by replacing the leftmost \
-non-overlapping occurrences o...''')
+        tiptest(p.sub, '(repl, string, count=0)')
```

Comment:
tiptest calls idlelib.calltip.get_argspec(p.sub).  The latter creates a tip from p.sub's signature and __doc__ attribute.  I verified that with this patch, p.sub.__doc__ is None and that help(p.sub) ends with the sentence "Return the string obtained by replacing the leftmost non-overlapping occurrences of pattern in string by the replacement repl.".  This sentence is p.sub.__doc__ in installed 3.10.0a2.

Help (pydoc) gets the docstring from `re.Pattern.sub.__doc__`.  IDLE no longer does, because `p.sub.__doc__` no longer gets looked-up on class function where it resides.  For a user class, `instance.method.__doc__` is `class.method.__doc__`.

This patch changes `type(p.sub)` from `builtin function or method` to `builtin function`.  Does that have anything to do with the behavior change?  It seems like a bug.

Reply:
Yes, this is pretty much what I did in a6bc4810d6eb4551d192db90b6779f96a1cf47ab, (which I reverted some minutes ago).

> Help (pydoc) gets the docstring from `re.Pattern.sub.__doc__`. IDLE no longer does, because `p.sub.__doc__` no longer gets looked-up on class function where it resides. For a user class, `instance.method.__doc__` is `class.method.__doc__`.

Aha; thanks
 
> This patch changes `type(p.sub)` from `builtin function or method` to `builtin function`. Does that have anything to do with the behavior change? It seems like a bug.

I believe it does. As I noted, the `dbm` module has similar issues. It definitely seems like a bug, and it seems to be tied to the `METH_METHOD` flag. I've tried looking in `Objects/descrobject.c` and `Objects/methodobject.c` without luck (I'm on thin ice in that directory).

Reply:
… or could it be a bug in `pydoc.py`?

Reply:
In `Objects/methodobject.c`, `meth_get__doc__()` is never called for `PyCMethod_Type` objects, only for `PyCFunction_Type` objects.

Reply:
@terryjreedy `re.Pattern.sub` has changed from `builtin function or method` to `builtin method` (as it is now defined with the `METH_METHOD` flag). I guess the bug is in IDLE, then? Should I apply a6bc481 again, and file an issue for IDLE?

Reply:
FYI, `inspect.getdoc` gets it right:
```
>>> import inspect
>>> import re
>>> p = re.compile('')
>>> inspect.getdoc(p.sub)
'Return the string obtained by replacing the leftmost non-overlapping occurrences of pattern in string by the replacement repl.'
>>> inspect.getdoc(re.Pattern.sub)
'Return the string obtained by replacing the leftmost non-overlapping occurrences of pattern in string by the replacement repl.'
```

Reply:
This fixes call tips in IDLE, but the REPL is still unable to fetch the docstring:
```
diff --git a/Lib/idlelib/calltip.py b/Lib/idlelib/calltip.py
index 549e224015..87a963dc1f 100644
--- a/Lib/idlelib/calltip.py
+++ b/Lib/idlelib/calltip.py
@@ -185,10 +185,7 @@ def get_argspec(ob):
              if len(argspec) > _MAX_COLS else [argspec] if argspec else [])
 
     # Augment lines from docstring, if any, and join to get argspec.
-    if isinstance(ob_call, types.MethodType):
-        doc = ob_call.__doc__
-    else:
-        doc = getattr(ob, "__doc__", "")
+    doc = inspect.getdoc(ob)
     if doc:
         for line in doc.split('\n', _MAX_LINES)[:_MAX_LINES]:
             line = line.strip()
```

Reply:
Thank your for pointing out getdoc.  It turns out that special-casing MethodType for getting the docstring is not needed for any existing test.  Using inspect.getdoc is the right way to go, as it gets docstrings more often.  In fact, it breaks two other tests by getting docstrings that are currently missed.   I switched to getdoc and adjusted those tests in new issue 42415, #23416.  After I merge when tests pass, update your master and update this with `git merge upstream/master`.  Then tiptest(p.sub,...) should pass.

Reply:
> Thank your for pointing out getdoc. It turns out that special-casing MethodType for getting the docstring is not needed for any existing test. Using inspect.getdoc is the right way to go, as it gets docstrings more often. In fact, it breaks two other tests by getting docstrings that are currently missed. I switched to getdoc and adjusted those tests in new issue 42415, #23416. After I merge when tests pass, update your master and update this with `git merge upstream/master`. Then tiptest(p.sub,...) should pass.

Great, thanks!